File: agents/nats-executor/ssh/executor.go
package ssh

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/melbahja/goph"
	"github.com/nats-io/nats.go"
	"golang.org/x/crypto/ssh"
	"log"
	"nats-executor/local"
	"nats-executor/utils"
	"time"
)

func Execute(req ExecuteRequest, instanceId string) ExecuteResponse {
	auth := goph.Password(req.Password)
	client, err := goph.NewConn(&goph.Config{
		User:     req.User,
		Addr:     req.Host,
		Port:     22,
		Auth:     auth,
		Timeout:  30 * time.Second,
		Callback: ssh.InsecureIgnoreHostKey(), // 👈 跳过 known_hosts 验证
	})

	if err != nil {
		log.Printf("Failed to create new SSH client: %v", err)
		return ExecuteResponse{
			InstanceId: instanceId,
			Success:    false,
			Output:     fmt.Sprintf("Failed to create new SSH client: %v", err),
		}
	}

	defer client.Close()

	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(req.ExecuteTimeout)*time.Second)
	defer cancel()

	out, err := client.RunContext(ctx, req.Command)
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			log.Printf("Command timed out")
		} else {
			log.Printf("Command execution error: %v", err)
		}
		return ExecuteResponse{
			Output:     string(out),
			InstanceId: instanceId,
			Success:    false,
		}
	}

	return ExecuteResponse{
		Output:     string(out),
		InstanceId: instanceId,
		Success:    true,
	}
}

func SubscribeSSHExecutor(nc *nats.Conn, instanceId *string) {
	subject := fmt.Sprintf("ssh.execute.%s", *instanceId)
	log.Printf("Subscribing to subject: %s", subject)
	nc.Subscribe(subject, func(msg *nats.Msg) {
		// 解析 request 的标准结构
		var incoming struct {
			Args   []json.RawMessage      `json:"args"`
			Kwargs map[string]interface{} `json:"kwargs"`
		}

		if err := json.Unmarshal(msg.Data, &incoming); err != nil {
			log.Printf("Error unmarshalling incoming message: %v", err)
			return
		}

		if len(incoming.Args) == 0 {
			log.Printf("No arguments received")
			return
		}

		var sshExecuteRequest ExecuteRequest
		if err := json.Unmarshal(incoming.Args[0], &sshExecuteRequest); err != nil {
			log.Printf("Error unmarshalling first arg to ssh.ExecuteRequest: %v", err)
			return
		}

		log.Printf("Received SSH command: %s", sshExecuteRequest.Command)
		responseData := Execute(sshExecuteRequest, *instanceId)
		log.Printf("Publishing SSH response using msg.Respond")

		responseContent, _ := json.Marshal(responseData)
		if err := msg.Respond(responseContent); err != nil {
			log.Printf("Error responding to SSH request: %v", err)
		}
	})
}

func SubscribeDownloadToRemote(nc *nats.Conn, instanceId *string) {
	subject := fmt.Sprintf("download.remote.%s", *instanceId)
	log.Printf("Subscribing to subject: %s", subject)

	nc.Subscribe(subject, func(msg *nats.Msg) {
		var incoming struct {
			Args   []json.RawMessage      `json:"args"`
			Kwargs map[string]interface{} `json:"kwargs"`
		}

		if err := json.Unmarshal(msg.Data, &incoming); err != nil {
			log.Printf("Error unmarshalling incoming message: %v", err)
			return
		}

		if len(incoming.Args) == 0 {
			log.Printf("No arguments received")
			return
		}

		var downloadRequest DownloadFileRequest

		if err := json.Unmarshal(incoming.Args[0], &downloadRequest); err != nil {
			log.Printf("Error unmarshalling first arg to DownloadFileRequest: %v", err)
			return
		}

		log.Printf("Starting download from bucket %s, file %s to local path %s", downloadRequest.BucketName, downloadRequest.FileKey, downloadRequest.TargetPath)

		// 下载文件到本地
		localdownloadRequest := utils.DownloadFileRequest{
			BucketName:     downloadRequest.BucketName,
			FileKey:        downloadRequest.FileKey,
			FileName:       downloadRequest.FileName,
			TargetPath:     downloadRequest.TargetPath,
			ExecuteTimeout: downloadRequest.ExecuteTimeout,
		}

		err := utils.DownloadFile(localdownloadRequest, nc)
		if err != nil {
			log.Printf("Error downloading file: %v", err)
			return
		}

		// 使用sshpass处理带密码的scp传输
		scpCommand := fmt.Sprintf("sshpass -p '%s' scp -o StrictHostKeyChecking=no %s/%s %s@%s:%s",
			downloadRequest.Password,
			localdownloadRequest.TargetPath,
			localdownloadRequest.FileName,
			downloadRequest.User,
			downloadRequest.Host,
			downloadRequest.TargetPath)

		localExecuteRequest := local.ExecuteRequest{
			Command:        scpCommand,
			ExecuteTimeout: downloadRequest.ExecuteTimeout,
		}

		log.Printf("Starting file transfer to remote host: %s@%s:%s", downloadRequest.User, downloadRequest.Host, downloadRequest.TargetPath)
		responseData := local.Execute(localExecuteRequest, *instanceId)

		responseContent, _ := json.Marshal(responseData)
		if err := msg.Respond(responseContent); err != nil {
			log.Printf("Error responding to download request: %v", err)
		}
	})
}

func SubscribeUploadToRemote(nc *nats.Conn, instanceId *string) {
	subject := fmt.Sprintf("upload.remote.%s", *instanceId)
	log.Printf("Subscribing to subject: %s", subject)

	nc.Subscribe(subject, func(msg *nats.Msg) {
		var incoming struct {
			Args   []json.RawMessage      `json:"args"`
			Kwargs map[string]interface{} `json:"kwargs"`
		}

		if err := json.Unmarshal(msg.Data, &incoming); err != nil {
			log.Printf("Error unmarshalling incoming message: %v", err)
			return
		}

		if len(incoming.Args) == 0 {
			log.Printf("No arguments received")
			return
		}

		var uploadRequest UploadFileRequest

		if err := json.Unmarshal(incoming.Args[0], &uploadRequest); err != nil {
			log.Printf("Error unmarshalling first arg to UploadFileRequest: %v", err)
			return
		}

		log.Printf("Starting upload from local path %s to remote host %s@%s:%s", uploadRequest.SourcePath, uploadRequest.User, uploadRequest.Host, uploadRequest.TargetPath)

		// 使用sshpass处理带密码的scp传输
		scpCommand := fmt.Sprintf("sshpass -p '%s' scp -o StrictHostKeyChecking=no %s %s@%s:%s",
			uploadRequest.Password,
			uploadRequest.SourcePath,
			uploadRequest.User,
			uploadRequest.Host,
			uploadRequest.TargetPath)

		localExecuteRequest := local.ExecuteRequest{
			Command:        scpCommand,
			ExecuteTimeout: uploadRequest.ExecuteTimeout,
		}

		log.Printf("Executing SCP command to upload file")
		responseData := local.Execute(localExecuteRequest, *instanceId)

		responseContent, _ := json.Marshal(responseData)
		if err := msg.Respond(responseContent); err != nil {
			log.Printf("Error responding to upload request: %v", err)
		}
	})
}

File: server/apps/rpc/executor.py
from apps.rpc.base import RpcClient


class Executor(object):
    def __init__(self, instance_id):
        """
        命令执行客户端
        :param instance_id: 执行器实例ID
        """
        self.instance_id = instance_id
        self.local_client = RpcClient('local.execute')
        self.ssh_client = RpcClient('ssh.execute')
        self.download_to_local_client = RpcClient('download.local')
        self.download_to_remote_client = RpcClient('download.remote')
        self.transfer_file_to_remote_client = RpcClient('upload.remote')
        self.unzip_local_client = RpcClient('unzip.local')

    def execute_local(self, command, timeout=60):
        """
        执行本地命令
        :param command: 要执行的命令
        :param timeout: 执行超时时间(秒)
        :return: 命令执行结果
        """
        request_data = {
            "command": command,
            "execute_timeout": timeout
        }
        return_data = self.local_client.run(self.instance_id, request_data, _timeout=timeout)
        return return_data

    def execute_ssh(self, command, host, username, password=None, key_file=None, timeout=60, port=22):
        """
        通过SSH执行远程命令
        :param command: 要执行的命令
        :param host: 远程主机地址
        :param port: 远程主机端口(可选)
        :param username: SSH用户名
        :param password: SSH密码(可选)
        :param key_file: SSH密钥文件路径(可选)
        :param timeout: 执行超时时间(秒)
        :return: 命令执行结果
        """
        request_data = {
            "command": command,
            "host": host,
            "port": port,
            "user": username,
            "execute_timeout": timeout
        }

        # 添加可选参数
        if password:
            request_data["password"] = password
        if key_file:
            request_data["key_file"] = key_file

        return_data = self.ssh_client.run(self.instance_id, request_data, _timeout=timeout)
        return return_data

    def download_to_local(self, bucket_name, file_key, file_name, target_path, timeout=60):
        """
        下载文件
        :param bucket_name: 存储桶名称
        :param file_key: 文件在存储桶中的键
        :param file_name: 文件名称
        :param target_path: 本地目标路径
        :param timeout: 执行超时时间(秒)
        :return: 下载结果
        """
        request_data = {
            "bucket_name": bucket_name,
            "file_key": file_key,
            "file_name": file_name,
            "target_path": target_path,
            "execute_timeout": timeout
        }
        return_data = self.download_to_local_client.run(self.instance_id, request_data, _timeout=timeout)
        return return_data

    def download_to_remote(self, bucket_name, file_key, file_name, target_path, host, username, password=None, timeout=60, port=22):
        """
        下载文件到远程
        :param bucket_name: 存储桶名称
        :param file_key: 文件在存储桶中的键
        :param file_name: 文件名称
        :param target_path: 远程目标路径
        :param host: 远程主机地址
        :param port: 远程主机端口(可选)
        :param username: SSH用户名
        :param password: SSH密码(可选)
        :param timeout: 执行超时时间(秒)
        :return: 下载结果
        """
        request_data = {
            "bucket_name": bucket_name,
            "file_key": file_key,
            "file_name": file_name,
            "target_path": target_path,
            "host": host,
            "port": port,
            "user": username,
            "execute_timeout": timeout,
        }
        # 添加可选参数
        if password:
            request_data["password"] = password
        return_data = self.download_to_remote_client.run(self.instance_id, request_data, _timeout=timeout)
        return return_data

    def unzip_local(self, file_path, target_path, timeout=60):
        """
        解压本地文件
        :param file_path: 要解压的文件路径
        :param target_path: 解压目标路径
        :return: 解压结果
        """
        request_data = {
            "zip_path": file_path,
            "dest_dir": target_path,
        }
        return_data = self.unzip_local_client.run(self.instance_id, request_data, _timeout=timeout)
        return return_data

    def transfer_file_to_remote(self, source_path, target_path, host, username, password=None, timeout=60, port=22):
        """
        传递文件到远程主机
        :param source_path: 要传递的文件路径
        :param target_path: 远程目标路径
        :param host: 远程主机地址
        :param port: 远程主机端口(可选)
        :param username: SSH用户名
        :param password: SSH密码(可选)
        :param timeout: 执行超时时间(秒)
        :return: 传递结果
        """
        request_data = {
            "source_path": source_path,
            "target_path": target_path,
            "host": host,
            "port": port,
            "user": username,
            "execute_timeout": timeout,
        }
        # 添加可选参数
        if password:
            request_data["password"] = password
        return_data = self.transfer_file_to_remote_client.run(self.instance_id, request_data, _timeout=timeout)
        return return_data
